sqlalchemy-migrate -> Alembic

smotornyuk opened the pull request:
Use alembic instead of sqlalchemy-migrate. Main reason - I just like it. Second reason - it's more flexible and supports branching in an easy way. Third reason - [sqlalchemy-migrate has some problems with python>=3.6](https://github.com/openstack/sqlalchemy-migrate/pull/12)

I rewrote all migrations in alembic so that we can safely drop sqlalchemy-migrate before switching to Py3. Theoretically, there will be no problem for new installations, where alembic is used to create the database from scratch. For existing portals, I'm going to use Alembic for all migrations that were added in CKAN>=2.5(starting from 82). That's mean, that portals that use CKAN<2.5 must use a three-step upgrade workflow:

1.  upgrade codebase to some version >=2.5 and <2.9 
2. Upgrade database
3. Then upgrade both codebase and database to the desired version as per docs

As for CLI, I'm going to leave all commands as is and update commands internally, so there is no need in updating a lot of documentation or rewriting any of the existing installation scripts.

Are there any objections or suggestions?


PS Just for fun I'm also writing  downgrade scripts, even thou I'm not sure whether they have any value

amercader commented on ckan/migration/alembic.ini:
```diff
@@ -0,0 +1,74 @@
+# are written from script.py.mako
+# output_encoding = utf-8
+
+sqlalchemy.url = postgresql://ckan_default:root@localhost/ckan_alembic
```
Is this DB actually used? is it a placeholder?

smotornyuk replied:
Nope, this is default configuration file benerated durinig alembic initialization, that can be used for development purposes and debuging(for example, you can run all alembic commands directly from folder, containing alembic.ini), but CKAN commands(paster, etc.) are initialized with database url from main configuration file. Probably, i should set `user:password` and `dbname` to some base values, like ones from initial configuration file. Or this line may be removed at all...

amercader commented on ckan/migration/__init__.py:
```diff
@@ -0,0 +1,8 @@
+# encoding: utf-8
+
+
+def skip_based_on_legacy_engine_version(op, filename):
```
Can you explain what this function does or what is needed for?

wardi replied:
just a docsting would be nice e.g. "graceful upgrade from old migration numbers to alembic ids"

wardi commented:
This looks good to me. The handling and deletion of the sqlalchemy migrate version is nice.

wardi commented on ckan/migration/versions/088_3ad397f70903_remove_migrate_version_table.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def downgrade():
+    pass
```
maybe add a comment about creating migrate_version not being necessary once we've switched to alembic ids

smotornyuk replied:
Yep, good idea

smotornyuk commented:
@amercader this solution should be compatible with existing instances starting from ckan-2.5. I'm checking in every migration existing of sqlalcemy-migrate's version table via `skip_based_on_legacy_engine_version` - i'll add docstring to it, but it basically compares prefix of migration file(001, 032, etc) with current version pulled from `migrate_version` table(if latter exists). If there is no such table, all alembic migrations will be applied, otherwise - as long as alembic migration is before the current version of sqlalchemy-migration, it applied without actual execution of migration script(this done via check in first two lines of `upgrade` functions). Theoretically it may work for any version of existing portals, but i tested it only for range between 2.5 and master.

davidread commented:
Another reason for moving to Alembic is because sqlalchemy-migrate is 'not maintained': https://groups.google.com/forum/#!topic/sqlalchemy/eF1DE20FMCQ

wardi commented:
I think this just needs conflicts resolved

smotornyuk commented:
@wardi , i've updated PR. There are two new migrations that i've converted to alembic migrations and small changes in CLI commands for database management

wardi commented on ckan/cli/db.py:
```diff
@@ -56,17 +57,49 @@ def updatedb(version=None):
+@db.command(u'downgrade', short_help=u'Downgrade the database')
+@click.option(u'-v', u'--version', help=u'Migration version', default=u'base')
+def downgradedb(version):
+    u'''Upgrading the database'''
```
Upgrading?

davidread replied:
😄 
```suggestion
    u'''Downgrading the database'''
```

smotornyuk replied:
Thank you:)
